File: tests/support/lxgb_metadata.py
import hashlib

VECTOR_ASSET = bytes(range(0, 32))
VECTOR_ISSUER_PUBLIC = bytes(range(32, 64))
VECTOR_SALT = bytes(range(64, 96))
VECTOR_LENGTH = 439
VECTOR_SHA256 = '38d5d09e3fc241bd6b6f924e4fe656f48954a61055ddec048f1eb636a72c8fe4'
SYMBOL_MAX = 16
DECIMALS_MAX = 38
DEFAULT_SYMBOL = b'TST'
DEFAULT_DECIMALS = 6


def custody_reference(asset):
    reference = bytes(12) + asset[12:]
    if not any(reference):
        raise ValueError('paxeer custody reference must be non-zero')
    return reference


def asset_symbol(symbol):
    if isinstance(symbol, str):
        symbol = symbol.encode('ascii')
    if not 0 < len(symbol) <= SYMBOL_MAX or any(byte > 0x7f for byte in symbol):
        raise ValueError('asset symbol must be 1 to %d ASCII bytes' % SYMBOL_MAX)
    return bytes(symbol)


def asset_decimals(decimals):
    if type(decimals) is not int or not 0 <= decimals <= DECIMALS_MAX:
        raise ValueError('asset decimals must be an integer in 0..%d' % DECIMALS_MAX)
    return decimals


def metadata(asset, issuer_public, salt, symbol=DEFAULT_SYMBOL, decimals=DEFAULT_DECIMALS):
    if len(asset) != 32 or len(issuer_public) != 32 or len(salt) != 32:
        raise ValueError('asset, issuer public key and supplied salt must be 32 bytes')
    symbol = asset_symbol(symbol)
    decimals = asset_decimals(decimals)
    did = ('did:layerx:' + issuer_public.hex()).encode()
    issuer = hashlib.sha256(b'LXP/v1/did-id\0' + len(did).to_bytes(2, 'big') + did).digest()
    reference = custody_reference(asset)
    record = (b'\0\x03' + asset + len(symbol).to_bytes(1, 'big') + symbol + decimals.to_bytes(1, 'big')
              + b'\x02' + len(reference).to_bytes(2, 'big') + reference
              + b'\0\x0dCustody token' + bytes(16) + issuer + b'\x02' + bytes(16) + salt)
    schedule = b'\0\x02' + bytes(80) + (10000).to_bytes(4, 'big') + b'\x0a'
    schedule += b''.join((0).to_bytes(16, 'big') for _ in range(10))
    return b'\0\x01' + len(record).to_bytes(2, 'big') + record + len(schedule).to_bytes(2, 'big') + schedule


def metadata_withdrawal(asset, issuer_public, salt, withdrawal_price=0,
                        symbol=DEFAULT_SYMBOL, decimals=DEFAULT_DECIMALS):
    if type(withdrawal_price) is not int or not 0 <= withdrawal_price < 2**64:
        raise ValueError('withdrawal fee price must be a u64')
    legacy = metadata(asset, issuer_public, salt, symbol, decimals)
    schedule = bytearray(legacy[-247:])
    schedule[:2] = b'\0\x03'
    schedule[86] = 11
    schedule += withdrawal_price.to_bytes(8, 'big')
    return legacy[:-249] + len(schedule).to_bytes(2, 'big') + schedule


def metadata_modules(asset, issuer_public, salt, withdrawal_price, prices,
                     symbol=DEFAULT_SYMBOL, decimals=DEFAULT_DECIMALS):
    if len(prices) != 7 or any(type(price) is not int or not 0 <= price < 2**128 for price in prices):
        raise ValueError('exactly seven uint128 module fee prices are required')
    legacy = metadata_withdrawal(asset, issuer_public, salt, withdrawal_price, symbol, decimals)
    schedule = b'\0\x04' + legacy[-253:] + b'\x07'
    schedule += b''.join(price.to_bytes(16, 'big') for price in prices)
    return legacy[:-257] + len(schedule).to_bytes(2, 'big') + schedule


def check():
    encoded = metadata(VECTOR_ASSET, VECTOR_ISSUER_PUBLIC, VECTOR_SALT)
    digest = hashlib.sha256(encoded).hexdigest()
    if len(encoded) != VECTOR_LENGTH or digest != VECTOR_SHA256:
        raise AssertionError('lxgb metadata emitter drift: length=%d sha256=%s; '
                             'tests/support/lxgb_metadata.rs pins the same vector'
                             % (len(encoded), digest))
    return encoded


if __name__ == '__main__':
    check()
    print('lxgb metadata vector %s length %d' % (VECTOR_SHA256, VECTOR_LENGTH))

File: tests/support/lxgb_metadata.rs
use sha2::{Digest, Sha256};
use std::fmt::Write as _;

fn length(value: usize) -> [u8; 2] {
    u16::try_from(value)
        .unwrap_or_else(|error| panic!("metadata field length: {error:?}"))
        .to_be_bytes()
}

pub fn custody_reference(asset: &[u8; 32]) -> [u8; 32] {
    let mut reference = [0_u8; 32];
    reference[12..].copy_from_slice(&asset[12..]);
    assert!(
        reference.iter().any(|byte| *byte != 0),
        "paxeer custody reference must be non-zero"
    );
    reference
}

pub fn append(request: &mut Vec<u8>, asset: &[u8; 32], issuer_public: &[u8; 32], salt: &[u8; 32]) {
    let mut public_hex = String::with_capacity(64);
    for byte in issuer_public {
        write!(public_hex, "{byte:02x}")
            .unwrap_or_else(|error| panic!("issuer public key encoding: {error:?}"));
    }
    let did = format!("did:layerx:{public_hex}");
    let mut digest = Sha256::new();
    digest.update(b"LXP/v1/did-id\0");
    digest.update(length(did.len()));
    digest.update(did.as_bytes());
    let issuer: [u8; 32] = digest.finalize().into();
    let reference = custody_reference(asset);
    let mut record = 3_u16.to_be_bytes().to_vec();
    record.extend_from_slice(asset);
    record.extend_from_slice(b"\x03TST\x06\x02");
    record.extend_from_slice(&length(reference.len()));
    record.extend_from_slice(&reference);
    record.extend_from_slice(b"\x00\x0dCustody token");
    record.extend_from_slice(&0_u128.to_be_bytes());
    record.extend_from_slice(&issuer);
    record.push(2);
    record.extend_from_slice(&0_u128.to_be_bytes());
    record.extend_from_slice(salt);
    assert_eq!(record.len(), 186);
    request.extend_from_slice(&1_u16.to_be_bytes());
    request.extend_from_slice(&length(record.len()));
    request.extend_from_slice(&record);
    let mut schedule = 2_u16.to_be_bytes().to_vec();
    for value in [0_u128; 5] {
        schedule.extend_from_slice(&value.to_be_bytes());
    }
    schedule.extend_from_slice(&10000_u32.to_be_bytes());
    schedule.push(10);
    for value in [0_u128; 10] {
        schedule.extend_from_slice(&value.to_be_bytes());
    }
    assert_eq!(schedule.len(), 247);
    request.extend_from_slice(&length(schedule.len()));
    request.extend_from_slice(&schedule);
}

pub fn append_withdrawal(
    request: &mut Vec<u8>,
    asset: &[u8; 32],
    issuer_public: &[u8; 32],
    salt: &[u8; 32],
    withdrawal_price: u64,
) {
    append(request, asset, issuer_public, salt);
    let schedule = request.len() - 247;
    request[schedule - 2..schedule].copy_from_slice(&255_u16.to_be_bytes());
    request[schedule..schedule + 2].copy_from_slice(&3_u16.to_be_bytes());
    request[schedule + 86] = 11;
    request.extend_from_slice(&withdrawal_price.to_be_bytes());
}

#[cfg(test)]
mod tests {
    use super::append;
    use sha2::{Digest, Sha256};

    const VECTOR_LENGTH: usize = 439;
    const VECTOR_SHA256: &str = "38d5d09e3fc241bd6b6f924e4fe656f48954a61055ddec048f1eb636a72c8fe4";

    fn sequence(start: u8) -> [u8; 32] {
        let mut bytes = [0_u8; 32];
        let mut value = start;
        for byte in &mut bytes {
            *byte = value;
            value = value.wrapping_add(1);
        }
        bytes
    }

    #[test]
    fn withdrawal_metadata_matches_the_native_fee_encoder() {
        let mut legacy = Vec::new();
        append(&mut legacy, &sequence(0), &sequence(32), &sequence(64));
        let mut encoded = Vec::new();
        super::append_withdrawal(&mut encoded, &sequence(0), &sequence(32), &sequence(64), 17);
        assert_eq!(encoded.len(), VECTOR_LENGTH + 8);
        assert_eq!(&encoded[..190], &legacy[..190]);
        assert_eq!(&encoded[190..192], &255_u16.to_be_bytes());
        assert_eq!(
            &encoded[192..],
            include_bytes!("../fixtures/fee-params-v3.bin")
        );
    }

    #[test]
    fn emits_the_shared_vector() {
        let mut encoded = Vec::new();
        append(&mut encoded, &sequence(0), &sequence(32), &sequence(64));
        assert_eq!(encoded.len(), VECTOR_LENGTH, "lxgb metadata vector length");
        let digest: [u8; 32] = Sha256::digest(&encoded).into();
        let mut expected = [0_u8; 32];
        for (index, byte) in expected.iter_mut().enumerate() {
            *byte = u8::from_str_radix(&VECTOR_SHA256[index * 2..index * 2 + 2], 16)
                .unwrap_or_else(|error| panic!("pinned vector digest: {error:?}"));
        }
        assert_eq!(
            digest, expected,
            "lxgb metadata emitters disagree; tests/support/lxgb_metadata.py pins the same vector"
        );
    }
}
